### PR TITLE
feat: improve performance with rpc query

### PR DESCRIPTION
**Pull request by Pythonberg1997**
### Description

This pr is to improve performance with rpc query

### Changes

Notable changes:
* remove `baseapp.qms` and add `queryState` for query
* add `baseapp.checkStateMtx` and `baseapp.queryStateMtx`

### CODE REVIEW

File: baseapp/baseapp.go [hunk 1]
```diff
@@ -80,11 +80,16 @@ type BaseApp struct { // nolint: maligned
...
 	// checkState is set on InitChain and reset on Commit
 	// deliverState is set on InitChain and BeginBlock and set to nil on Commit
+	// queryState is set on InitChain and BeginBlock
+	queryState           *state // optional alternative multistore for querying only.
 	checkState           *state // for CheckTx
 	deliverState         *state // for DeliverTx
 	processProposalState *state // for ProcessProposal
 	prepareProposalState *state // for PrepareProposal
 
+	queryStateMtx *sync.RWMutex // mutex for queryState
```
Comment (guagualvcha) [inline baseapp/baseapp.go:90]: No need to be pointer

File: baseapp/baseapp.go [hunk 2]
```diff
@@ -179,6 +184,16 @@ func NewBaseApp(
 		app.SetMempool(mempool.NoOpMempool{})
 	}
 
+	if app.queryStateMtx == nil {
```
Comment (guagualvcha) [inline baseapp/baseapp.go:187]: no need to new if it is not pointer.

File: baseapp/baseapp.go [hunk 3]
```diff
@@ -578,10 +608,20 @@ func (app *BaseApp) getState(mode runTxMode) *state {
...
 
 	default:
+		app.queryStateMtx.RLock()
+		defer app.queryStateMtx.RUnlock()
+
 		return app.checkState
 	}
 }
 
+func (app *BaseApp) getQueryState() *state {
```
Comment (guagualvcha) [inline baseapp/baseapp.go:618]: Please modify the `checkTx`, as if it change check state, we should lock it.

Comment (guagualvcha) [reply baseapp/baseapp.go:618]: I think the commit function change the checkstate as well, we should lock there

File: baseapp/baseapp.go [hunk 4]
```diff
@@ -578,10 +608,20 @@ func (app *BaseApp) getState(mode runTxMode) *state {
 		return app.processProposalState
```
Comment (guagualvcha) [inline baseapp/baseapp.go:609]: I hope `handleEthQuery` can use the query state as well.

File: baseapp/abci.go [hunk 5]
```diff
@@ -474,7 +474,9 @@ func (app *BaseApp) Commit() abci.ResponseCommit {
 	// The write to the DeliverTx state writes all state transitions to the root
 	// MultiStore (app.cms) so when Commit() is called is persists those values.
 	app.deliverState.ms.Write()
+	app.queryStateMtx.Lock()
```
Comment (guagualvcha) [inline baseapp/abci.go:477]: It is possible not necessary.

File: baseapp/baseapp.go [hunk 6]
```diff
@@ -458,6 +468,19 @@ func (app *BaseApp) setState(mode runTxMode, header tmproto.Header) {
 	}
 }
 
+func (app *BaseApp) setQueryState(header tmproto.Header) {
+	var ms sdk.CommitMultiStore
+	if _, ok := app.cms.(*rootmulti.Store); ok {
```
Comment (guagualvcha) [inline baseapp/baseapp.go:473]: make it simple:
```
	if mstore, ok := app.cms.(*rootmulti.Store); ok {
		ms = mstore.DeepCopy()
	}
```

File: baseapp/abci.go [hunk 7]
```diff
@@ -194,6 +195,18 @@ func (app *BaseApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseBeg
 			WithBlockHeight(req.Header.Height)
 	}
 
+	app.queryStateMtx.Lock()
+	if app.queryState == nil {
+		app.setQueryState(req.Header)
```
Comment (guagualvcha) [inline baseapp/abci.go:200]: we should do:
`app.setQueryState(req.Header)`
 
in `commit` as well, once a block is commit, the query state should be updated.

File: baseapp/abci.go [hunk 8]
```diff
@@ -840,7 +871,15 @@ func (app *BaseApp) CreateQueryContext(height int64, prove bool) (sdk.Context, e
 			)
 	}
 
-	cacheMS, err := qms.CacheMultiStoreWithVersion(height)
+	var cacheMS storetypes.CacheMultiStore
+	var err error
+	if len(path) == 1 && path[0] == "/cosmos.auth.v1beta1.Query/Account" && height == lastBlockHeight {
+		// use checkState for account queries on the latest height
+		// we could get the newest account info to send multi txs in one block
+		cacheMS = app.checkState.ms
```
Comment (guagualvcha) [inline baseapp/abci.go:879]: do you think a checkState read lock is required here?

**Review by guagualvcha — APPROVED** [review]

**Review by KeefeL — APPROVED** [review]

**Pull request merged**